File: ClockKotlin/src/main/kotlin/au/id/wale/swin/clock/Clock.kt
package au.id.wale.swin.clock

class Clock {
    private var seconds = Counter()
    private var minutes = Counter()
    private var hours = Counter()

    val currentTime: String
        get() {
            val formattedHours = "%02d".format(hours.count)
            val formattedMinutes = "%02d".format(minutes.count)
            val formattedSeconds = "%02d".format(seconds.count)

            return "${formattedHours}:${formattedMinutes}:${formattedSeconds}"
        }

    fun reset() {
        seconds.reset()
        minutes.reset()
        hours.reset()
    }

    fun tick() {
        if(seconds.count < 59) {
            seconds.increment()
        } else {
            seconds.reset()
            if(minutes.count < 59) {
                minutes.increment()
            } else {
                minutes.reset()
                if (hours.count >= 23) {
                    hours.reset()
                } else {
                    hours.increment()
                }
            }
        }
    }
}

File: ClockApp/ClockApp/Clock.cs
namespace ClockApp;
public class Clock
{
    private Counter _seconds, _minutes, _hours;
    public Clock()
    {
        _seconds = new Counter();
        _minutes = new Counter();
        _hours = new Counter();
    }

    public string CurrentTime 
    {
        get {
            return $"{_hours.Count.ToString("00")}:{_minutes.Count.ToString("00")}:{_seconds.Count.ToString("00")}";
        }
    }

    public void Reset()
    {
        _seconds.Reset();
        _minutes.Reset();
        _hours.Reset();
    }

    public void Tick()
    {
        if (_seconds.Count < 59)
        {
            _seconds.Increment();
        }
        else
        {
            _seconds.Reset();
            if (_minutes.Count < 59)
            {
                _minutes.Increment();
            }
            else
            {
                _minutes.Reset();
                if (_hours.Count > 23)
                {
                    _hours.Reset();
                }
                else
                {
                    _hours.Increment();
                }
            }
        }
    }
}